### PR TITLE
Dashrews/ROX-14606 ROX-14630 ROX-15202 ROX-15240 partition network flows

### PR DESCRIPTION
**Pull request by dashrews78**
## Description

A lot of the performance issues we have been experiencing with network flows is related to our access pattern.  Our stores for network flows are based on the cluster.  All our queries utilize the cluster.  When we have a large amount of data in a single table we get into situations where if a significant chunk of the data is returned for a given cluster then Postgres will decide it is more efficient to run a full table scan.  This access pattern lends itself to partitioning the table based on the cluster.  This way Postgres will direct the queries to the partition for that cluster.  So even if we are doing a full table scan we are only doing a scan of the data for that specific cluster vs all of the data.  This greatly improves the performance of the queries as they are operating on significantly less data and it is easier for Postgres to keep the statistics up to date as each partition has its only statistics.

Now Gorm does not support partitions so network flows can no longer be managed by Gorm.

Additionally some of the indexes were btree where hash indexes are going to perform better for how we are querying the data so those have been updated in this PR as well.

It is important to note here, that this change is more for longer term stability and performance improvements.  The improvements will not show up until there is a significant load and number for flows in the system.  At which point the single network_flows table starts to degrade as it falls behind in stats and simply has hot clusters that consume enough of the table that it is more efficient to scan the entire table to find the necessary rows.

Some data:
1. Query to simply get the current flows for a cluster:
```
central_active=# explain analyze SELECT nf.Props_SrcEntity_Type, nf.Props_SrcEntity_Id, nf.Props_DstEntity_Type, nf.Props_DstEntity_Id, nf.Props_DstPort, nf.Props_L4Protocol, nf.LastSeenTimestamp, nf.ClusterId::text FROM network_flows nf  INNER JOIN (SELECT Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId, MAX(Flow_Id) AS MaxFlow FROM network_flows GROUP BY Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId) tmpflow on nf.Props_SrcEntity_Type = tmpflow.Props_SrcEntity_Type AND nf.Props_SrcEntity_Id = tmpflow.Props_SrcEntity_Id AND nf.Props_DstEntity_Type = tmpflow.Props_DstEntity_Type AND nf.Props_DstEntity_Id = tmpflow.Props_DstEntity_Id AND nf.Props_DstPort = tmpflow.Props_DstPort AND nf.Props_L4Protocol = tmpflow.Props_L4Protocol AND nf.ClusterId = tmpflow.ClusterId and nf.Flow_id = tmpflow.MaxFlow  WHERE nf.ClusterId = '13ba7c72-6ac4-4497-99ab-5fae27125972';
                                                                                                                                                                                                                                   QUERY PLAN               
                                                                                                                                                                                                                    
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=1772136.66..2003733.13 rows=1 width=130) (actual time=143268.459..145978.854 rows=442247 loops=1)
   Hash Cond: ((network_flows.props_srcentity_type = nf.props_srcentity_type) AND ((network_flows.props_srcentity_id)::text = (nf.props_srcentity_id)::text) AND (network_flows.props_dstentity_type = nf.props_dstentity_type) AND ((network_flows.props_ds
tentity_id)::text = (nf.props_dstentity_id)::text) AND (network_flows.props_dstport = nf.props_dstport) AND (network_flows.props_l4protocol = nf.props_l4protocol) AND ((max(network_flows.flow_id)) = nf.flow_id))
   ->  Finalize GroupAggregate  (cost=871438.65..1031635.22 rows=771485 width=114) (actual time=3419.357..4798.898 rows=442247 loops=1)
         Group Key: network_flows.props_srcentity_type, network_flows.props_srcentity_id, network_flows.props_dstentity_type, network_flows.props_dstentity_id, network_flows.props_dstport, network_flows.props_l4protocol, network_flows.clusterid
         ->  Gather Merge  (cost=871438.65..1003813.97 rows=1005320 width=114) (actual time=3419.325..4465.043 rows=715462 loops=1)
               Workers Planned: 2
               Workers Launched: 2
               ->  Partial GroupAggregate  (cost=870438.63..886775.08 rows=502660 width=114) (actual time=3398.851..3993.847 rows=238487 loops=3)
                     Group Key: network_flows.props_srcentity_type, network_flows.props_srcentity_id, network_flows.props_dstentity_type, network_flows.props_dstentity_id, network_flows.props_dstport, network_flows.props_l4protocol, network_flows.clust
erid
                     ->  Sort  (cost=870438.63..871695.28 rows=502660 width=114) (actual time=3398.835..3813.065 rows=276748 loops=3)
                           Sort Key: network_flows.props_srcentity_type, network_flows.props_srcentity_id, network_flows.props_dstentity_type, network_flows.props_dstentity_id, network_flows.props_dstport, network_flows.props_l4protocol
                           Sort Method: external merge  Disk: 33448kB
                           Worker 0:  Sort Method: external merge  Disk: 36976kB
                           Worker 1:  Sort Method: external merge  Disk: 35336kB
                           ->  Parallel Seq Scan on network_flows  (cost=0.00..760986.68 rows=502660 width=114) (actual time=0.354..2477.526 rows=276748 loops=3)
                                 Filter: (clusterid = '13ba7c72-6ac4-4497-99ab-5fae27125972'::uuid)
                                 Rows Removed by Filter: 2578573
   ->  Hash  (cost=845137.43..845137.43 rows=1206385 width=122) (actual time=139848.290..139848.291 rows=830245 loops=1)
         Buckets: 32768  Batches: 64  Memory Usage: 2288kB
         ->  Seq Scan on network_flows nf  (cost=0.00..845137.43 rows=1206385 width=122) (actual time=2.431..138173.338 rows=830245 loops=1)
               Filter: (clusterid = '13ba7c72-6ac4-4497-99ab-5fae27125972'::uuid)
               Rows Removed by Filter: 7735720
 Planning Time: 0.529 ms
 Execution Time: 146005.271 ms
(24 rows)
```

Partitioned plan.  Note the plan here is basically the same as above, the only difference is that the partitioned plan is working on a subset of the data where as the plan above is working on all of it.:
```
central_active=# explain analyze SELECT nf.Props_SrcEntity_Type, nf.Props_SrcEntity_Id, nf.Props_DstEntity_Type, nf.Props_DstEntity_Id, nf.Props_DstPort, nf.Props_L4Protocol, nf.LastSeenTimestamp, nf.ClusterId::text FROM network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe nf  INNER JOIN (SELECT Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId, MAX(Flow_Id) AS MaxFlow FROM network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe GROUP BY Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId) tmpflow on nf.Props_SrcEntity_Type = tmpflow.Props_SrcEntity_Type AND nf.Props_SrcEntity_Id = tmpflow.Props_SrcEntity_Id AND nf.Props_DstEntity_Type = tmpflow.Props_DstEntity_Type AND nf.Props_DstEntity_Id = tmpflow.Props_DstEntity_Id AND nf.Props_DstPort = tmpflow.Props_DstPort AND nf.Props_L4Protocol = tmpflow.Props_L4Protocol AND nf.ClusterId = tmpflow.ClusterId and nf.Flow_id = tmpflow.MaxFlow  WHERE nf.ClusterId = '598ec5a0-d086-4bbf-990b-213072bf12fe';
                                                                                                                                                                                                                                                            
                                                                                                                   QUERY PLAN                                                                                                                               
                                                                                                                                                                                                                                                
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=275810.83..357367.70 rows=1 width=130) (actual time=7137.521..14069.606 rows=915586 loops=1)
   Hash Cond: ((network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_type = nf.props_srcentity_type) AND ((network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_id)::text = (nf.props_srcentity_id)::text) AND (network_fl
ows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_type = nf.props_dstentity_type) AND ((network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_id)::text = (nf.props_dstentity_id)::text) AND (network_flows_v2_598ec5a0_d086_4bbf_
990b_213072bf12fe.props_dstport = nf.props_dstport) AND (network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_l4protocol = nf.props_l4protocol) AND ((max(network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.flow_id)) = nf.flow_id))
   ->  Finalize GroupAggregate  (cost=163980.19..211712.47 rows=124109 width=135) (actual time=2753.289..5967.261 rows=915586 loops=1)
         Group Key: network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_type, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_type, networ
k_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstport, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_l4protocol, network_flows_v2_598ec5a0_d086_4bbf_990b_21
3072bf12fe.clusterid
         ->  Gather Merge  (cost=163980.19..205507.02 rows=248218 width=135) (actual time=2753.274..5049.368 rows=1137341 loops=1)
               Workers Planned: 2
               Workers Launched: 2
               ->  Partial GroupAggregate  (cost=162980.17..175856.48 rows=124109 width=135) (actual time=2688.743..4026.117 rows=379114 loops=3)
                     Group Key: network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_type, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_
type, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstport, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_l4protocol, network_flows_v2_598ec5a0_d086_
4bbf_990b_213072bf12fe.clusterid
                     ->  Sort  (cost=162980.17..164272.97 rows=517121 width=135) (actual time=2688.721..3565.548 rows=413781 loops=3)
                           Sort Key: network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_type, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_srcentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dsten
tity_type, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstentity_id, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_dstport, network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe.props_l4protocol
                           Sort Method: external merge  Disk: 62000kB
                           Worker 0:  Sort Method: external merge  Disk: 64056kB
                           Worker 1:  Sort Method: external merge  Disk: 61112kB
                           ->  Parallel Seq Scan on network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe  (cost=0.00..43198.02 rows=517121 width=135) (actual time=0.021..244.802 rows=413781 loops=3)
                                 Filter: ((clusterid)::text = '598ec5a0-d086-4bbf-990b-213072bf12fe'::text)
   ->  Hash  (cost=52247.64..52247.64 rows=1241091 width=143) (actual time=4381.624..4381.625 rows=1241344 loops=1)
         Buckets: 32768  Batches: 64  Memory Usage: 3744kB
         ->  Seq Scan on network_flows_v2_598ec5a0_d086_4bbf_990b_213072bf12fe nf  (cost=0.00..52247.64 rows=1241091 width=143) (actual time=0.172..1735.795 rows=1241344 loops=1)
               Filter: ((clusterid)::text = '598ec5a0-d086-4bbf-990b-213072bf12fe'::text)
 Planning Time: 2.009 ms
 Execution Time: 14148.481 ms
(22 rows)
```


2. Old query plan.  Notice the sequential scan that had to filter 28M rows.  This is due to the fact that at a certain point the query on cluster will return a large enough percentage of the table that Postgres will just scan the entire thing.
```
central_active=# explain analyze select * from network_flows where clusterid = 'cec45863-4e13-4686-b6c8-f9781240a5a8';
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Seq Scan on network_flows  (cost=0.00..1071586.35 rows=3370489 width=114) (actual time=0.016..6421.039 rows=3184714 loops=1)
   Filter: (clusterid = 'cec45863-4e13-4686-b6c8-f9781240a5a8'::uuid)
   Rows Removed by Filter: 28847751
 Planning Time: 0.109 ms
 Execution Time: 6538.668 ms
(5 rows)
```

With partitions. Notice it is a full sequence scan, but JUST on the partition:
```
explain analyze select * from network_flows_v2 where clusterid = '1a66b8c3-1967-4507-ba04-f9185cbe4767';
                                                                                    QUERY PLAN                                                                                    
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Seq Scan on network_flows_v2_1a66b8c3_1967_4507_ba04_f9185cbe4767 network_flows_v2  (cost=0.00..33507.40 rows=840432 width=143) (actual time=0.037..316.707 rows=918898 loops=1)
   Filter: ((clusterid)::text = '1a66b8c3-1967-4507-ba04-f9185cbe4767'::text)
 Planning Time: 0.195 ms
 Execution Time: 350.615 ms
(4 rows)
```

Now for the Remove Stale Query.
Old plan:
```
```

Partition plan:
```
explain analyze DELETE FROM network_flows_5b6c0b80_56aa_4618_b6c7_2735242471b3 a USING (
      SELECT MAX(flow_id) as Max_Flow, Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId
        FROM network_flows      
WHERE ClusterId = '5b6c0b80-56aa-4618-b6c7-2735242471b3' GROUP BY Props_SrcEntity_Type, Props_SrcEntity_Id, Props_DstEntity_Type, Props_DstEntity_Id, Props_DstPort, Props_L4Protocol, ClusterId
HAVING COUNT(*) > 1
      ) b
      WHERE a.Props_SrcEntity_Type = b.Props_SrcEntity_Type
 AND a.Props_SrcEntity_Id = b.Props_SrcEntity_Id
 AND a.Props_DstEntity_Type = b.Props_DstEntity_Type
 AND a.Props_DstEntity_Id = b.Props_DstEntity_Id
AND a.Props_DstPort = b.Props_DstPort
AND a.Props_L4Protocol = b.Props_L4Protocol
AND a.ClusterId = b.ClusterId
      AND a.Flow_Id <> b.Max_Flow;
                                                                                                                                                                                               QUERY PLAN
                                                                                                                                         
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on network_flows_5b6c0b80_56aa_4618_b6c7_2735242471b3 a  (cost=559704.82..751990.80 rows=1 width=165) (actual time=18267.084..18267.089 rows=0 loops=1)
   ->  Hash Join  (cost=559704.82..751990.80 rows=1 width=165) (actual time=6882.652..12732.651 rows=940516 loops=1)
         Hash Cond: ((a.props_srcentity_type = b.props_srcentity_type) AND ((a.props_srcentity_id)::text = (b.props_srcentity_id)::text) AND (a.props_dstentity_type = b.props_dstentity_type) AND ((a.props_dstentity_id)::text = (b.props_dstent
ity_id)::text) AND (a.props_dstport = b.props_dstport) AND (a.props_l4protocol = b.props_l4protocol) AND ((a.clusterid)::text = (b.clusterid)::text))
         Join Filter: (a.flow_id <> b.max_flow)
         Rows Removed by Join Filter: 940516
         ->  Seq Scan on network_flows_5b6c0b80_56aa_4618_b6c7_2735242471b3 a  (cost=0.00..70709.36 rows=2003136 width=141) (actual time=0.022..1207.863 rows=2080192 loops=1)
         ->  Hash  (cost=555259.61..555259.61 rows=66771 width=294) (actual time=6880.377..6880.380 rows=940516 loops=1)
               Buckets: 16384 (originally 16384)  Batches: 128 (originally 8)  Memory Usage: 3969kB
               ->  Subquery Scan on b  (cost=473840.48..555259.61 rows=66771 width=294) (actual time=3416.508..5963.209 rows=940516 loops=1)
                     ->  HashAggregate  (cost=473840.48..554591.91 rows=66771 width=135) (actual time=3416.007..5657.285 rows=940516 loops=1)
                           Group Key: network_flows.props_srcentity_type, network_flows.props_srcentity_id, network_flows.props_dstentity_type, network_flows.props_dstentity_id, network_flows.props_dstport, network_flows.props_l4protocol, net
work_flows.clusterid
                           Filter: (count(*) > 1)
                           Planned Partitions: 16  Batches: 265  Memory Usage: 4121kB  Disk Usage: 342448kB
                           Rows Removed by Filter: 199160
                           ->  Seq Scan on network_flows_5b6c0b80_56aa_4618_b6c7_2735242471b3 network_flows  (cost=0.00..75717.20 rows=2003136 width=135) (actual time=0.005..1439.755 rows=2080192 loops=1)
                                 Filter: ((clusterid)::text = '5b6c0b80-56aa-4618-b6c7-2735242471b3'::text)
 Planning Time: 0.505 ms
 Execution Time: 18344.206 ms
(18 rows)
```

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

No real changes to the primary tests as the whole point is that those should work as before only faster.  A migration was required and tests were created for that.

Additionally scale runs were used repeatedly to narrow in on how best to utilize the partitions for best performance.

### CODE REVIEW

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (dashrews78) [inline central/networkgraph/flow/datastore/internal/store/postgres/store.go:150]: This works inserting directly into the parent, BUT it is recommended to insert into the partition or use triggers to do so.  In our case, it is pretty simple to just use the partition directly.

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (connorgorman) [inline central/networkgraph/flow/datastore/internal/store/postgres/store.go:60]: Wonder if we can pass in the clusterid to access the partition faster? Would that actually matter? Maybe the planner knows that the clusterid scopes the partition?

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (connorgorman) [inline central/networkgraph/flow/datastore/internal/store/postgres/store.go:214]: This may be unnecessary, but I'm wary of having %s in the queries. Can we validate that cluster id is a uuid and ensure that someone couldn't create a sql injection in a cluster id?

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (connorgorman) [inline central/networkgraph/flow/datastore/internal/store/postgres/store.go:285]: obviously remove this before merging

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (dashrews78) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go:214]: verifying that is a UUID first is a good idea.  the other option is to just generate a new id for the partition, but using the cluster seemed to make sense.

File: migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go
Comment (connorgorman) [inline migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go:108]: Tempted to just pull this from the clusters table instead?

**Review by connorgorman — CHANGES_REQUESTED** [review]
See comments

File: migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go
Comment (dashrews78) [reply migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go:108]: oh that is a good idea.  That will be much faster than doing the way I did.

File: migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go
Comment (dashrews78) [reply migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go:138]: yeah I wrote a follow up ticket and started that effort on a separate branch.  I don't want to hold this one up for that so depends on timing as whether I put it in now or immediately after.

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (dashrews78) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go:60]: It is supposed to be able to figure that out and seems to do so fairly well.  The places where I had to help it and point it to the partition were on inserts as that is recommended and when we had to do those joins.

File: migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go
Comment (dashrews78) [reply migrator/migrations/m_172_to_m_173_network_flows_partition/migration.go:138]: and my inclination is to do the timeout subsequently because I have to get a good idea for what to use.

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (connorgorman) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go:214]: 👍

**Review by connorgorman — APPROVED** [review]

**Pull request merged**
